spec/classes/lib/ccreds_spec.rb

```diff
@@ -1,146 +1,198 @@
+
+  params = {
+    'package_ensure' =>
+      { :value => 'installed',
+        :type  => 'string',
+      },
+    'package_name' =>
+      { :value => 'USE_DEFAULTS',
+        :type  => 'string',
+      },
+  }
+
+  describe 'with an invalid value for parameter' do
+    params.sort.each do |k,v|
+      context "#{k}" do
+        let :facts do
+          { :osfamily                  => 'RedHat',
+            :operatingsystemmajrelease => '7',
           }
         end
-      it { should contain_package('pam_ccreds').with(
-        'ensure'  => 'absent'
-      ) }
-    end
-    context 'when given a custom package' do
-      let :params do
-          {
-            :package  => 'libpam-magic-pkg'
-          }
-        end
-      it { should contain_package('libpam-magic-pkg') }
-    end
-  end
-  context 'on a Ubuntu OS' do
-    let(:facts) {
-      {
-        :osfamily       => 'Debian',
-        :lsbdistrelease => '12.04',
-        :lsbdistid      => 'Ubuntu',
-      }
-    }
-    context 'with no parameters' do
-      it { should contain_package('libpam-ccreds').with(
-        'ensure'  => 'installed'
-      ) }
-    end
-    context 'when absent' do
-      let :params do
-          {
-            :ensure  => 'absent'
-          }
+
+        case v[:type]
+        when 'array'
+          kvalue = 'invalid_type'
+          errormsg = '"invalid_type" is not an array.  It looks to be a String.'
+        when 'string'
+          kvalue = ['invalid','type']
+          errormsg = '\["invalid", "type"\] is not a string.  It looks to be a Array'
+        when 'path'
+          kvalue = 'invalid/path'
+          errormsg = '"invalid/path" is not an absolute path.'
+        when 'bool'
+          kvalue = ['invalid','type']
+          errormsg = '\["invalid", "type"\] is not a boolean.  It looks to be a Array'
+        when 'hash'
+          kvalue = ['invalid','type']
+          errormsg = '\["invalid", "type"\] is not a hash.  It looks to be a Array'
+        else
+          raise "error in your spec tests - you have an unknown type value (#{v[:type]}) for parameter #{k}."
         end
-      it { should contain_package('libpam-ccreds').with(
-        'ensure'  => 'absent'
-      ) }
-    end
-    context 'when given a custom package' do
-      let :params do
-          {
-            :package  => 'libpam-magic-pkg'
-          }
+
+        let(:params) { { :"#{k}" => kvalue } }
+
+        it 'should fail' do
+          expect {
+            should contain_class('pam::lib::ccreds')
+          }.to raise_error(Puppet::Error,/#{errormsg}/)
         end
-      it { should contain_package('libpam-magic-pkg') }
+      end
     end
   end
-  context 'on a Solaris OS' do
-    let(:facts) {
-      {
-        :osfamily      => 'Solaris',
-        :kernelrelease => '5.11',
+
+  describe 'with a valid custom value specified for parameter' do
+    let :facts do
+      { :osfamily                  => 'RedHat',
+        :operatingsystemmajrelease => '7',
```

Comment:
We are testing that if you pass a value to a param that something happens. We aren't testing that it would be different based on the platform. That would happen in the first bit of testing. This is just saying, I passed a value and here's what I expect. Because of this, the platform doesnt really matter, it just needs to be valid.